common/trace/src/trace.rs

```diff
@@ -390,9 +388,23 @@ fn abbreviate_crate(name: &str) -> &str {
     }
 }
 
-/// Sets diem trace config
+/// Sets diem trace config.
+///
+/// This should only be called once.
 pub fn set_diem_trace(config: &HashMap<String, String>) -> Result<()> {
-    SAMPLING_CONFIG.get_or_try_init(|| parse_sampling_config(config))?;
+    // Ensure that this function is called just once. OnceCell guarantees that its initializer is
+    // called exactly once.
+    let mut initializer_called = false;
+    SAMPLING_CONFIG
+        .get_or_try_init(|| {
+            initializer_called = true;
+            parse_sampling_config(config)
+        })
+        .with_context(|| "failed to parse sampling config")?;
+
+    if !initializer_called {
+        bail!("failed to initialize: set_diem_trace called multiple times")
+    }
```

Comment:
```suggestion
let config = parse_sampling_config(config)?;
SAMPLING_CONFG.set(confg).context("failed to initialize: set_diem_trace called multiple times")
```

Something like this maybe? of course you're solution would also work

Reply:
Looking at https://docs.rs/once_cell/1.5.2/src/once_cell/lib.rs.html#794-801 I think they're basically equivalent but I still prefer my solution slightly because it means we don't have to pay the cost of parsing the sampling config if we know it's going to fail. /shrug